contract-message-test/tests/harness.rs

```diff
@@ -1,184 +1,672 @@
+            &wallet,
             TxParameters::default(),
+            StorageConfiguration::default(),
+            Salt::from_str(RANDOM_SALT).unwrap(),
         )
         .await
         .unwrap();
-    let _receipt = wallet
-        .transfer(
-            &wallet.address(),
-            transfer_amount,
-            native_asset,
+
+        // Create optional inputs/outputs
+        let input_contract = Input::Contract {
+            utxo_id: UtxoId::new(Bytes32::zeroed(), 0u8),
+            balance_root: Bytes32::zeroed(),
+            state_root: Bytes32::zeroed(),
+            tx_pointer: TxPointer::default(),
+            contract_id: other_contract_id.into(),
+        };
+        let output_variable1 = Output::Variable {
+            to: Address::default(),
+            amount: Word::default(),
+            asset_id: AssetId::default(),
+        };
+        let output_variable2 = Output::Variable {
+            to: Address::default(),
+            amount: Word::default(),
+            asset_id: AssetId::from_str(RANDOM_SALT).unwrap(),
+        };
+        let output_contract = Output::Contract {
+            input_index: 3u8,
+            balance_root: Bytes32::zeroed(),
+            state_root: Bytes32::zeroed(),
+        };
+        let output_message = Output::Message {
+            recipient: Address::default(),
+            amount: Word::default(),
+        };
+
+        // Relay the test message to the test contract
+        let _receipts = env::relay_message_to_contract(
+            &wallet,
+            message_inputs[0].clone(),
+            contract_input,
+            &coin_inputs[..],
+            &vec![input_contract],
+            &vec![
+                output_variable1,
+                output_variable2,
+                output_contract,
+                output_message,
+            ],
+        )
+        .await;
+
+        // Verify test contract received the message
+        let test_contract_counter = test_contract.get_test_counter().call().await.unwrap().value;
+        assert_eq!(test_contract_counter, 1);
+
+        // Verify test contract received the correct data
+        let test_contract_id: ContractId = test_contract._get_contract_id().into();
+        let test_contract_data1 = test_contract.get_test_data1().call().await.unwrap().value;
+        assert_eq!(test_contract_data1, test_contract_id);
+
+        // Verify the message value was received by the test contract
+        let provider = wallet.get_provider().unwrap();
+        let test_contract_balance = provider
+            .get_contract_asset_balance(test_contract._get_contract_id(), AssetId::default())
+            .await
+            .unwrap();
+        assert_eq!(test_contract_balance, 420);
+    }
+}
+
+// Test the cases where the transaction should panic due to the
+// predicate script failing to validate the transaction requirements
+mod panic {
+    use std::str::FromStr;
+
+    use crate::utils::builder;
```

Comment:
I would take a look at which imports are being used where, similar to variables etc. and refactor them out e.g. builder is used here and in the top level.

Reply:
I'm not sure what you mean. Modules don't auto inherit things at the top level. `use utils::builder;` is defined at the top level to ensure it's compiled and accessible via `crate::` in the modules. I am realizing now that the `const` defined at the top are no longer being used (from before the different pass and fail modules)